docs/proposals/affinity.md

```diff
@@ -0,0 +1,341 @@
+S have this HardAffinityExpression in their PodSpec, then the HardAffinityExpression will block the first
+pod of the service from ever scheduling, since it is only allowed to run in a zone with another pod from
+the same service. And of course that means none of the pods of the service will be able
+to schedule. This problem *only* applies to hard affinity, not soft affinity or hard or soft
+anti-affinity. There are at least two ways to solve this problem
+* short-term: when a controller creates pods from a PodTemplate, for exactly *one* of those
+pods, it should omit any hard affinity rules that select the pods of that PodTemplate.
+* long-term/speculative: controllers could present the scheduler with a group of pods from
+the same PodTemplate as a single unit. This would allow two things: (1) the scheduler could ignore
+problematic hard affinity rules rather than requiring the controller to handle it,
+and (2) the scheduler could do proper [gang scheduling](https://github.com/kubernetes/kubernetes/issues/16845)
+since it could receive an entire gang simultaneously.
+
+### Anti-affinity
+
+* **Spread the pods of this service S across nodes and zones**:
+`{{PodSelector: <selector that matches S's pods>, TopologyKey: "node"}, {PodSelector: <selector that matches S's pods>, TopologyKey: "zone"}}`
+
+* **Don't co-locate pods of this service with pods from service "evilService"**:
+`{PodSelector: selector that matches evilService's pods, TopologyKey: "node"}`
+
+* **Don't co-locate pods of this service with any other pods**:
+`{PodSelector: empty, TopologyKey: "node"}`
+
+* **Don't co-locate pods of this service with any other pods except other pods of this service**:
+Assuming pods from the service have some label `{key=service, value=S}`:
+`{PodSelector: "service" NotIn "S", TopologyKey: "node"}`
+Note that this works because `"service" NotIn "S"` matches pods with no key "service"
+as well as pods with key "service" and a corresponding value that is not "S."
+
+## Algorithm
+
+An example algorithm a scheduler might use to implement affinity and anti-affinity rules is as follows.
+There are certainly more efficient ways to do it; this is just intended to demonstrate that the API's
+semantics are implementable.
+
+Terminology definition: We say a pod P is "feasible" on a node N if P meets all of the scheduler
+predicates for scheduling P onto N. Note that there are various reasons why a pod that is correctly
+scheduled onto some node might become infeasible on that node later; for example,
+some labels might have changed on the node (thus violating some constraint of an already-running pod)
+or some labels might have changed on a pod (thus violating some AffinitySelector of an already-running pod).
+We assume pods that become infeasible after they are bound to a node may be allowed to continue running; the
+various policies and mechanisms for handling this scenario are outside the scope of this document, and
```

Comment:
As mentioned above, I think that hard constraints are of limited to no practical use until we answer the above conundrum.  Lets do that before finalizing the API or implementing hard constraints.